### PR TITLE
A tool for determing max rps of a backend or proxy

### PR DESCRIPTION
**Pull request by stevej**
Problem:
Determining whether a branch has an adverse effect on system
throughput is currently done by hand in an error-prone fashion.

Solution:
This is an automatic tool based on the Universal Scalability Law
that runs a load test at a few levels of concurrency and determines
the max concurrency (meaning at which point extra concurrency slows
response time) and max requests per second of the strest-grpc backend
or intermediary. The only thing you might tune is the concurrency levels
at very high scale.

Verification:
Testing this experimentally has matched expectations.

### CODE REVIEW

**Comment by stevej** [comment]
Running this on our performance host, I notice that max rps is like 20% off from what I would expect. Let me work on this a bit more before anybody commits time to a review.

**Comment by stevej** [comment]
Ok, you can't reuse a grpc client without incurring some serious overhead at higher concurrency levels. Now it experimentally matches what I see with `strest-client`

File: max-rps/main.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+		log.Fatalf("timePerLevel cannot be less than 1 second.")
+	}
+
+	levels := strings.Split(*concurrencyLevels, ",")
+	var denseLatency [](float64)
+
+	for _, l := range levels {
+		level, err := strconv.Atoi(l)
+		if err != nil {
+			log.Fatalf("unknown concurrency level: %s", l, err)
```
Comment (klingerf) [inline max-rps/main.go]: Looks like you're missing a second placeholder in your string here. I think it should be:

```
log.Fatalf("unknown concurrency level: %s, %s", l, err)
```

Or similar.

Comment (stevej) [inline max-rps/main.go]: Thanks, I've fixed that.

**Pull request merged**